packages/synapse-react-client/src/components/EntityDownloadButton/EntityDownloadButton.tsx

```diff
@@ -70,55 +142,186 @@ export function getDownloadActionsForEntityType(
-    case 'link':
+    case EntityType.link:
       return [[DownloadAction.programmaticAccess]]
     default:
-      // This will fail if a new EntityType is added and not handled
+      // this will fail if a new EntityType is added and not handled
       throw new Error(`Unhandled EntityType: ${type}`)
   }
 }
 
+// get the appropriate version number for download based on entity type
+// for datasets: returns the most recent released version if any exist, undefined otherwise
+// for other entities: returns the current version number
+function useGetLatestVersionNumber(entityId: string, entityType: EntityType) {
+  // get entity data
+  const { data: entityData, isLoading: entityDataLoading } =
+    useGetEntity(entityId) // No version = latest
+
+  // for datasets, check if any versions exist and use the most recent one
+  const { data: versionsData, isLoading: versionsLoading } = useGetVersions(
+    entityId,
+    0,
+    50,
+    {
+      enabled: entityType === EntityType.dataset,
+    },
+  )
+
+  let latestVersionNumber: number | undefined
+
+  if (entityType === EntityType.dataset) {
+    // for datasets, check if any versions exist
+    if (versionsData?.results && versionsData.results.length > 0) {
+      // Use the most recent released version (first in the list, since versions are returned in descending order)
+      latestVersionNumber = versionsData.results[0].versionNumber
+    } else {
+      // no versions exist, use undefined (will use current/draft version)
+      latestVersionNumber = undefined
+    }
+  } else {
+    // for non-datasets, use the entity's version number
+    latestVersionNumber =
+      entityData && isVersionableEntity(entityData)
+        ? entityData.versionNumber
+        : undefined
+  }
+
+  const isLoading =
+    entityDataLoading ||
+    (entityType === EntityType.dataset ? versionsLoading : false)
+
+  return { latestVersionNumber, isLoading }
+}
+
+// create default queryBundleRequest with appropriate SQL
+function getDefaultQueryBundleRequestForEntity(
+  entityId: string,
+  versionNumber: number | undefined,
+  entityType: EntityType,
+): QueryBundleRequest {
+  let sql: string
+
+  if (entityType === EntityType.dataset) {
+    // for datasets:
+    // - if a version number exists (released version), use it: syn123.5
+    // - if no version number (no released versions), use current/draft: syn123
+    sql = versionNumber
+      ? `SELECT * FROM ${entityId}.${versionNumber}`
+      : `SELECT * FROM ${entityId}`
+  } else {
+    // For non-datasets, always use entity ID without version
+    sql = `SELECT * FROM ${entityId}`
+  }
+
+  return {
+    concreteType: 'org.sagebionetworks.repo.model.table.QueryBundleRequest',
+    entityId: `${entityId}`,
+    query: {
+      sql: sql,
+    } as Query,
+    partMask: 0,
+  }
+}
+
 export function EntityDownloadButton(props: {
   entityId: string
   name: string
   entityType: EntityType
```

Comment:
For this to be reusable, I think we would also need to accept `versionNumber` as a prop. Where we would have the following behavior:

- If the version number is present, use it
- If the version number is `null`, do not use it,
- If the version number is `undefined`, then we would try to do the 'smart' thing -- get the latest version for datasets, otherwise don't use it.

However, I think for the search scenario, this is fine. You may want to add a comment to this effect, if you agree.

Some of the utilities you have written may need to change accordingly (e.g. to separate the logical steps of 'getting a version if this is a dataset' and 'including the version in the instructions/query request'), if/when we make that change.